pkg/controller/direct/sql/normalize.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	if obj.Spec.RootPassword.ValueFrom != nil {
+		key := types.NamespacedName{
+			Namespace: obj.Namespace,
+			Name:      obj.Spec.RootPassword.ValueFrom.SecretKeyRef.Name,
+		}
+
+		secret := &corev1.Secret{}
+		if err := kube.Get(ctx, key, secret); err != nil {
+			if apierrors.IsNotFound(err) {
+				return k8s.NewSecretNotFoundError(key)
+			}
+			return fmt.Errorf("error reading referenced Secret %v: %w", key, err)
+		}
+
+		password := string(secret.Data[obj.Spec.RootPassword.ValueFrom.SecretKeyRef.Key])
+		obj.Spec.RootPassword.Value = &password
+
+		return nil
+	}
+	return nil
+}
+
+func normalizePrivateNetworkRef(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) error {
+	if obj.Spec.Settings.IpConfiguration == nil || obj.Spec.Settings.IpConfiguration.PrivateNetworkRef == nil {
+		return nil
+	}
+
+	resRef := obj.Spec.Settings.IpConfiguration.PrivateNetworkRef
+	netRef := &refs.ComputeNetworkRef{
+		External:  resRef.External,
+		Name:      resRef.Name,
+		Namespace: resRef.Namespace,
+	}
+	net, err := refs.ResolveComputeNetwork(ctx, kube, obj, netRef)
+	if err != nil {
+		return err
+	}
+
+	obj.Spec.Settings.IpConfiguration.PrivateNetworkRef.Name = net.String()
+
+	return nil
+}
+
+func normalizeAuditLogBucketRef(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) error {
+	if obj.Spec.Settings.SqlServerAuditConfig == nil {
+		return nil
+	}
+
+	if obj.Spec.Settings.SqlServerAuditConfig.BucketRef == nil {
+		return fmt.Errorf("must specify bucket for audit config")
+	}
+
+	bucketRef := obj.Spec.Settings.SqlServerAuditConfig.BucketRef
+
+	if bucketRef.External != "" && bucketRef.Name != "" {
+		return fmt.Errorf("cannot specify both spec.settings.sqlServerAuditConfig.bucketRef.external and spec.settings.sqlServerAuditConfig.bucketRef.name")
+	}
+
+	if bucketRef.External != "" {
+		bucketRef.Name = bucketRef.External
+	} else if bucketRef.Name != "" {
+		if bucketRef.Namespace == "" {
+			bucketRef.Namespace = obj.Namespace
+		}
+
+		key := types.NamespacedName{
+			Namespace: bucketRef.Namespace,
+			Name:      bucketRef.Name,
+		}
+
+		bucket := &unstructured.Unstructured{}
+		bucket.SetGroupVersionKind(storagev1beta1.StorageBucketGVK)
+		if err := kube.Get(ctx, key, bucket); err != nil {
+			if apierrors.IsNotFound(err) {
+				return k8s.NewReferenceNotFoundError(storagev1beta1.StorageBucketGVK, key)
+			}
+			return fmt.Errorf("error reading referenced StorageBucket %v: %w", key, err)
+		}
+
+		storageBucketName, err := refs.GetResourceID(bucket)
+		if err != nil {
+			return err
+		}
+		bucketRef.Name = "gs://" + storageBucketName
+	} else {
+		return fmt.Errorf("must specify either spec.settings.sqlServerAuditConfig.bucketRef.external or spec.settings.sqlServerAuditConfig.bucketRef.name")
+	}
+
+	return nil
+}
+
+func normalizeLabels(obj *krm.SQLInstance) error {
+	if obj.Labels == nil {
+		obj.Labels = make(map[string]string)
+	}
+	obj.Labels["managed-by-cnrm"] = "true"
```

Comment:
Just to confirm: `managed-by-cnrm` label is only persisted in the underlying GCP resource, but not in the KCC resource, right?

Reply:
Yes, that's correct